### PR TITLE
Features/checkout

### PR DESCRIPTION
**Pull request by aryansingh02**
descriptioin: fixed according to design. Fixed the comments. No styled components or old library present. 
verification: verified on both mobile and desktop screens manually

image
image

fixes https://github.com/SleekOrg/tracker/issues/294

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -13,26 +15,45 @@
...
     "@types/react": "^17.0.3",
+    "@types/react-bootstrap": "^0.32.25",
     "@types/react-dom": "^17.0.3",
+    "@types/react-global-configuration": "^1.3.1",
+    "@types/react-helmet": "^6.1.0",
+    "@types/react-icons": "^3.0.0",
+    "@types/react-phone-number-input": "^3.0.6",
     "@types/react-router-dom": "^5.1.7",
     "@types/redux": "^3.6.0",
+    "@types/styled-components": "^5.1.9",
```
Comment (teamSleek) [inline package.json:30]: Let us not migrate all the baggage from last app. Use modal equivalent in material UI and no custom styling as much as possible. If we migrate so much stuff then what is the point of making a new app?

**Review by teamSleek — CHANGES_REQUESTED** [review]

File: src/App.tsx [hunk 2]
```diff
@@ -1,15 +1,20 @@
 import React, { Component } from 'react';
 import './App.scss';
 import { Route, Switch, BrowserRouter as Router } from 'react-router-dom';
+import config from "react-global-configuration";
 import { SnackbarProvider } from 'notistack';
 import { WithStyles, withStyles, createStyles } from '@material-ui/core';
 import { theme as Theme } from './theme';
+import Cart from './features/cart/Cart';
```
Comment (teamSleek) [inline src/App.tsx:8]: Unable to visit /cart
Screen Shot 2021-04-06 at 2 39 32 PM

**Review by teamSleek — CHANGES_REQUESTED** [review]

**Comment by teamSleek** [comment]
Also, link the correct issue.

File: package.json (hunk 1 above)
Comment (aryansingh02) [reply package.json]: removed all stuff

**Comment by teamSleek** [comment]
Fix conflict

File: src/features/cart/Cart.tsx [hunk 3]
```diff
@@ -0,0 +1,420 @@
...
+  }
+
+  updateInvoice = async () => {
+    const { cart, stall } = this.props;
+    CalculateInvoice(cart).then((data) => {
+      this.setState({ invoice: data });
+    });
+  };
+
+  fetchOrderData = async () => {
```
Comment (teamSleek) [inline src/features/cart/Cart.tsx:188]: all this should be moved to the redux state

File: src/features/cart/Cart.tsx [hunk 4]
```diff
@@ -0,0 +1,420 @@
...
+
+    const response = await fetch(
+      `${config.get('backend')}/stall/${stall._id}/wait-time`
+    );
+    let time = await response.json();
+    time = GenerateWaitTime(time.waitTime);
+    this.setState({ itemEta: time });
+
+    if (dynamicEta < 1.0) {
+      this.triggerSkipTheLineModal();
```
Comment (teamSleek) [inline src/features/cart/Cart.tsx:200]: there is no modal in new design

File: src/features/cart/Cart.tsx [hunk 5]
```diff
@@ -0,0 +1,420 @@
+/* eslint no-param-reassign: 0 */ // --> OFF
```
Comment (teamSleek) [inline src/features/cart/Cart.tsx:1]: Delete all unused code. Copying and pasting from the previous app have brought in things that are not relevant in this app. Ideally we should NOT have copy-pasted and we should have coded things from scratch which would have kept things clean. 

https://sourcemaking.com/antipatterns/cut-and-paste-programming

File: src/utils.tsx [hunk 6]
```diff
@@ -42,4 +47,102 @@ export const GenerateExpenseLevel = (expenseLevel: number) => {
...
+      });
+    });
+  }
+  return {
+    amount: subTotal,
+    currency: cartItem.selectedVariation.price.currency,
+  };
+}
+
+export const GenerateCurrencyNumber = (price:IPrice) => (
```
Comment (teamSleek) [inline src/utils.tsx:103]: create a component for this.

File: src/utils.tsx [hunk 7]
```diff
@@ -42,4 +47,102 @@ export const GenerateExpenseLevel = (expenseLevel: number) => {
...
+export const GenerateEpochDate = (date:Date) => {
+  const year = date.getFullYear();
+  const month = date.getMonth();
+  const day = date.getDate();
+  // @ts-ignore
+  const newDate = Date.parse(new Date(year, month, day, 0, 0, 0, 0));
+  return newDate;
+};
+
+export async function CalculateInvoice(cart:ICart) {
```
Comment (teamSleek) [inline src/utils.tsx:131]: why do we even need this function?

File: src/utils.tsx [hunk 8]
```diff
@@ -42,4 +47,102 @@ export const GenerateExpenseLevel = (expenseLevel: number) => {
   return retVal;
 };
 
+export const DeleteLineItemFromCart = (cart:ICart, lineItemId:string) => {
```
Comment (teamSleek) [inline src/utils.tsx:50]: move all these cart management function in CartSlice or a CartUtil file

File: src/types.tsx [hunk 9]
```diff
@@ -1,21 +1,29 @@
...
 export type HistoryType = ReturnType<typeof Object>;
 
 export type IStall = typeof stall;
-
-export interface IPos {
-  type: string;
-  squareInfo: Record<string, string>;
-}
-
+export type ICart = typeof cart;
```
Comment (teamSleek) [inline src/types.tsx:9]: using ICart everywhere and defining it this way seems to be circumventing the entire Typescript checking itself. This was okay as long as we were doing UX and using this for mock data. Doing it for redux state management, API call, and all the data manipulation functions seems wrong.

File: src/features/event/Event.tsx [hunk 10]
```diff
@@ -130,5 +128,5 @@ class Event extends React.Component<IProps, IState> {
     );
   }
 }
-
-export default withWidth()(WithNavigation(withStyles(styles)(Event)));
+// @ts-ignore
```
Comment (teamSleek) [inline src/features/event/Event.tsx:131]: why do we need ts-ignore here? add comment

File: src/features/event/DesktopHeader.tsx [hunk 11]
```diff
@@ -126,7 +119,7 @@ export default function DesktopHeader() {
               aria-label="show more"
               aria-controls={mobileMenuId}
               aria-haspopup="true"
-              onClick={handleMobileMenuOpen}
+              // onClick={handleMobileMenuOpen}
```
Comment (teamSleek) [inline src/features/event/DesktopHeader.tsx:122]: Is this intended?

File: src/features/event/DesktopHeader.tsx [hunk 12]
```diff
@@ -86,16 +89,6 @@ const useStyles = makeStyles((theme: typeof Theme) =>
 
 export default function DesktopHeader() {
   const classes = useStyles();
-  const [anchorEl, setAnchorEl] = React.useState<null | HTMLElement>(null);
```
Comment (teamSleek) [inline src/features/event/DesktopHeader.tsx:89]: why delete all this?

File: src/features/cart/Cart.tsx [hunk 13]
```diff
@@ -0,0 +1,420 @@
...
+  constructor(props: IProps) {
+    super(props);
+    this.state = {
+      mode: FullfilmentModeType.FREE_PICKUP,
+      showModal: false,
+      dateModal: false,
+      itemEta: 0,
+      dynamicEta: 0,
+      dynamicFee: zeroFee,
+      invoice: undefined,
```
Comment (teamSleek) [inline src/features/cart/Cart.tsx:155]: Unable to run this.
Screen Shot 2021-04-09 at 2 35 26 PM

File: src/features/cart/Cart.tsx [hunk 14]
```diff
@@ -0,0 +1,420 @@
...
+      this.setState({ invoice: data });
+    });
+  };
+
+  fetchOrderData = async () => {
+    const { dynamicEta } = this.state;
+    const { stall } = this.props;
+
+    const response = await fetch(
+      `${config.get('backend')}/stall/${stall._id}/wait-time`
```
Comment (teamSleek) [inline src/features/cart/Cart.tsx:193]: where is this `backend` param defined ?

File: src/features/cart/Cart.tsx [hunk 15]
```diff
@@ -0,0 +1,420 @@
...
+    });
+  };
+
+  triggerSkipTheLineModal = async () => {
+    const { cart } = this.props;
+    if (!this.enableSurgeFee) {
+      return;
+    }
+
+    fetch(`${config.get('backend')}/surge-fee`, {
```
Comment (teamSleek) [inline src/features/cart/Cart.tsx:218]: ditto

**Review by teamSleek — CHANGES_REQUESTED** [review]

File: src/App.scss [hunk 16]
```diff
@@ -1,18 +1,39 @@
...
   align-items: center;
   justify-content: flex-start;
+  flex-direction: row;
+  flex-wrap: wrap;
 }
 
-body,
-html {
-  background-color: #f7f8fb;
+.centerJustifiedFlex {
```
Comment (teamSleek) [inline src/App.scss:15]: use Material grid and not custom css. Do not import it from the other app.

File: src/App.tsx [hunk 17]
```diff
@@ -1,16 +1,21 @@
...
+import configuration from './config';
 import Event from './features/event/Event';
 import { StallMenu } from './features/stall/StallMenu';
 import AddedToCart from './features/stall/AddedToCart';
 import OrderSent from './features/stall/OrderSent';
 import { OrderStatus } from './features/stall/OrderStatus';
 import { OrderConfirmation } from './features/stall/OrderConfirmation';
 
+config.set(configuration, { assign: false, freeze: false });
```
Comment (teamSleek) [inline src/App.tsx:17]: Tip section is missing.

File: src/features/cart/Cart.tsx [hunk 18]
```diff
@@ -0,0 +1,302 @@
...
+                />
+              </Grid>
+            )}
+            <Button
+              color="primary"
+              variant="contained"
+              className={classes.checkoutButton}
+              style={{ width: this.bodyWrapper?.current?.offsetWidth }}
+            >
+              <Typography roboto={true}>Proceed to Checkout</Typography>
```
Comment (teamSleek) [inline src/features/cart/Cart.tsx:255]: styling is off

**Review by teamSleek — CHANGES_REQUESTED** [review]
Styling very off in a lot of ways.

As part of the verification, step confirms that your thing matches with Figma. And screenshots of you app on desktop and mobile as proof.

File: .eslintrc [hunk 19]
```diff
@@ -92,7 +92,9 @@
     // the stateless function rule is unusually restrict
     "react/prefer-stateless-function": "off",
     // unusually retrict
-    "react/static-property-placement": "off"
+    "react/static-property-placement": "off",
```
[the remark this answers is not in the document]
Comment (teamSleek) [reply .eslintrc:95]: @aryansingh02 It seems that this still an open item?

**Review by teamSleek — CHANGES_REQUESTED** [review]
The styling and alignment seem to be off in the line items section of the cart.

File: src/features/event/DesktopHeader.tsx (hunk 12 above)
Comment (aryansingh02) [reply src/features/event/DesktopHeader.tsx]: was not needed

File: src/App.scss [hunk 20]
```diff
@@ -1,18 +1,40 @@
...
+}
+.boldFont{
+  font-weight: bold;
+}
+.centreAligned{
+  position: absolute;
+  top: 50%;
+  left: 50%;
+  transform: translate(-50%, -50%);
+}
```
[the remark this answers is not in the document]
Comment (aryansingh02) [reply src/App.scss]: it's not used in grid, used at some places to style boxes.

**Pull request merged**